### ISSUE TITLE
Timeout.InfiniteTimeSpan behaves counter-intuitively for some comparisons

### ISSUE THREAD

**Issue by invidious9000** [dotnet/runtime#37777]
### Description
Timeout.InfiniteTimeSpan uses an internal representation of -1ms to define a period of infinite time. The operators for <, >, >=, <= in TimeSpan compare ticks between two instances. This leads to comparisons that seem invalid when naively using InfiniteTimeSpan (negative ticks) on one side of the comparison.

Example:
```csharp
var oneSecond = TimeSpan.FromSeconds(1);
bool oneSecondIsLessThanInfinity = oneSecond < Timeout.InfiniteTimeSpan;
```
In this scenario `oneSecondIsLessThanInfinity` is false when one might expect it to be true.

.NET Core 3.1

**Comment by danmosemsft** [comment]
FWIW, same on .NET Framework.

**Comment by stephentoub** [comment]
I agree it's not intuitive, but there's very little that can be done about this. The meaning of the -1 value as a timeout is entirely up to how the consumer is defined and interprets it. Changing the comparison operations to special-case that value would also be very unintuitive, and a massive breaking change.

**Comment by invidious9000** [comment]
TimeSpan remarks indicate that the value contained is some count of ticks, and this can be enough information to reason about the behavior, but the comparison operators seem to be a slightly more opaque implementation detail (you must look at source to see why it behaves this way).

Could this be considered a doc issue? I see that kinda-similar concerns were identified/discussed in 
https://github.com/dotnet/runtime/issues/30723#issuecomment-574308350
>This really should be a doc issue demonstrating correct use of the existing Timeout API 

Is there any value in augmenting documentation for InfiniteTimeSpan to explicitly indicate nuances around comparison or is it reasonable to expect consumers to infer these from tick remarks? Would the same suggestions advocating for IntelliCode enhancements as outlined in https://github.com/dotnet/runtime/issues/30723#issuecomment-526676911 apply?

Could it be appropriate to add a see-also for TimeSpan.MaxValue if that can _sometimes_ be an acceptable substitute?

**Comment by stephentoub** [comment]
Thanks for the suggestions. Improving the docs is certainly possible! And we welcome contributions there, too.
https://github.com/dotnet/dotnet-api-docs

**Comment by julealgon** [comment]
> I agree it's not intuitive, but there's very little that can be done about this.

@stephentoub this could be a prime use case for discriminated unions when it is introduced.

**Comment by stephentoub** [comment]
> > I agree it's not intuitive, but there's very little that can be done about this.
> 
> [@stephentoub](https://github.com/stephentoub) this could be a prime use case for discriminated unions when it is introduced.

How so?

**Comment by stephentoub** [comment]
> > How so?
> 
> [@stephentoub](https://github.com/stephentoub) by isolating the "Infinite" value and decoupling it from using a negative `-1` integer internally: it would become a "special value" that could then have equality handled differently.

But how would that help with the existing API?

**Comment by stephentoub** [comment]
> That could also be cleanly resolved by having the "infinite" value be a discriminated union possibility instead of a magic "-1".

That would mean changing the type of Infinite{TimeSpan}, which won't happen.

**Comment by julealgon** [comment]
> That would mean changing the type of Infinite{TimeSpan}, **which won't happen.**

I would expect you to say that, but I hope you are (eventually) wrong. If the .NET team doesn't come up with a decent way to deprecate things over time the language will be overwhelmed by technical debt.

Yes, I know you did make _some_ very specific breaking changes, such as the ones around NRT and the `field` keyword, but that's not enough (IMHO). I hope the team realizes this before things get out of hand.

**Comment by CyrusNajmabadi** [comment]
Deprecating things causes debt as well.  You have to weigh the value you get out of the damage you cause.  In the vast majority of cases, things are working fine for the vast majority of users. So breaks would hurt the vast majority just to deal with a minor issue hit by a tiny set of users. 

This is not a good way to handle things.  Breaks are made when the value is there to justify it.

**Comment by julealgon** [comment]
> Deprecating things causes debt as well. You have to weigh the value you get out of the damage you cause.

This is true. But if done properly, that type of debt is temporary. 

Versus permanent tech debt that is created if you keep supporting the old thing forever.

There is certainly a trade-off to be made there.

> So breaks would hurt the vast majority just to deal with a minor issue hit by a tiny set of users.
> ...
> This is not a good way to handle things. Breaks are made when the value is there to justify it.

@CyrusNajmabadi I disagree with the approach because if you always follow that rule, you will end up in a situation where you are killed by a thousand cuts: each small thing "is too low-value to be considered in isolation", but now you have thousands of such small issues scattered throughout the .NET ecosystem and this makes the language/framework less intuitive, less consistent, etc, turning people off. I'm not saying .NET is already in that position, but I would argue that it will _invariably_ happen, if nothing is done about the problem conceptually.

Again, I don't want to derail this with this meta discussion. I just mentioned it here because this is one _clear_ case where discriminated unions would come in very handy to improve an existing abstraction, and this will keep happening more and more as the language evolves, to a point where (I think) you _will_ need to do something about it, if you want to keep the language ecosystem healthy.

If you'd rather restart from scratch and create a `C#++` at that point, then that's up to you... I hope that is not the course Microsoft wants to follow as it would massively fracture the community when it happens (much more so than a breaking change would).

**Comment by CyrusNajmabadi** [comment]
> but I would argue that it will invariably happen,

We're 25 years in. So far, everything is still going strong.  In practical terms we do not see that people "are killed by a thousand cuts" or are complaining that that's the case. 

> If you'd rather restart from scratch and create a C#++ at that point, then that's up to you... 

I think such a claim would need a ton of evidence to back it up. C# and .net did a ton to avoid the mistakes of c++. And 25 years in I think shows itself to be much healthier and in a much better than c++ is in.

**Comment by tannergooding** [comment]
Closing. This is by design and comes from using a sentinel value.
